### ISSUE TITLE
Boot hartid communication?

### ISSUE THREAD

**Issue by jrtc27** [riscv/riscv-acpi#5]
With OpenSBI and U-Boot currently making no guarantee about which hart is the boot hart that enters the kernel before the others are started via the SBI HSM extension, and mhartid being an M-mode CSR with no SBI interface to query it, what is the proposed strategy for ensuring an OS can match the boot hart with the list of harts in the ACPI tables and thus know which other harts it needs to start?

In the UEFI device tree world that information is passed in /chosen/boot-hartid, which was short-sighted as it should have just been available via a "get current hartid" SBI call (or the ISA made less awkward by defining an shartid). Either ACPI needs to grow an equivalent way to figure this out, or a better generic mechanism (SBI extension, ISA extension, or both) that obsoletes the device tree approach needs to be defined. The latter seems preferable, rather than furthering past mistakes.

**Comment by jrtc27** [comment]
Note that this issue doesn't exist on AArch64, where mpidr_el1 exists.

**Comment by vlsunil** [comment]
Thank you very much!, @jrtc27. 

I think we atleast need SBI extension which might be easier now than shartid. Parsing ACPI tables so early in EFI stub (linux) may not be easy since ACPI needs some level of memory management services to be setup. 

@atishp04, @avpatel - is there any reason to prefer DT method over an SBI extension? Can we add it if there are no major concerns?

**Comment by avpatel** [comment]
After the last M-mode booting stage (OpenSBI), only boot HART continues booting. The hartid of the booting hart is passed via a0 register. It is responsibility of S-mode software to save the hartid in some per-CPU structure.

Based on above approach, most S-mode software (such as, U-boot S-mode, Linux, Xvisor, etc) don't require special SBI extension for knowing boot hartid.

The boot-hartid being passed in DT by EDK2 is a work-around for EFI booting protocol. @atishp04 was there any other reason for passing boot-hartid via DT?

Regards,
Anup

**Comment by avpatel** [comment]
I vaguely recall that when booting kernel as EFI application there is no clean way of passing boot-hartid so rather than DT it will be good to pass boot-hartid as some kind of EFI parameter.

Regards,
Anup

**Comment by atishp04** [comment]
The /chosen/boot-hartid parameter was added for EFI stub to pass the hartid to proper kernel. It doesn't have any relation with EDK2.
EFI stub mimicks as a efi application. So it needs to know the boot hartid so that it can pass it to the proper kernel in a0.
IIRC, the other alternative discussed at that time was to define a RISC-V specific configuration table to describe the boot hartid. 
This will require to modify UEFI spec and an ECR which seemed an overkill at that point of time given the availability of DT was already available in the system tables.

I don't remember any other reason for choosing DT over the system configuration table. Henerich or Ard may remember. 
I would suggest to create a thread in efi and/or riscv mailing list to figure out a solution for ACPI.

**Comment by vlsunil** [comment]
Thanks @avpatel  and @atishp04 for some background details.

If I understand correctly, normal kernel should expect a0 to have the boot_hartid. Hence, this DT node is required only for booting Linux kernel as EFI application. For other OS which don't support booting as EFI application, passing boot_hartid via DT node is not required. Correct?  If that is the case, isn't it easier use volatile UEFI variable which is set by the EDK2/u-boot and read by the kernel if DT is not there?

**Comment by avpatel** [comment]
Any S-mode OS following the traditional RISC-V booting protocol don't need the boot-hartid DT property.

Only EFI stub based booting has no provision for passing boot-hartid so the boot-hartid DT property is a temporary work-around for EFI booting and this will break for UEFI+ACPI systems which do not have DT.

Ideally, the UEFI spec should define a RISC-V specific configuration parameter (or variable) so that boot-hartid can be discovered independent of UEFI+DT or UEFI+ACPI mechanism.

Atish was involved in related discussions so he might have pointers to past discussions.

Regards,
Anup

**Comment by jrtc27** [comment]
> After the last M-mode booting stage (OpenSBI), only boot HART continues booting. The hartid of the booting hart is passed via a0 register. It is responsibility of S-mode software to save the hartid in some per-CPU structure.
> 
> Based on above approach, most S-mode software (such as, U-boot S-mode, Linux, Xvisor, etc) don't require special SBI extension for knowing boot hartid.

_Most_ S-mode software should be booting via EFI, not directly from SBI firmware.

> The boot-hartid being passed in DT by EDK2 is a work-around for EFI booting protocol. @atishp04 was there any other reason for passing boot-hartid via DT?

I'm talking about how U-Boot behaves today when booting via EFI, not what the nascent EDK2 port does.

> If I understand correctly, normal kernel should expect a0 to have the boot_hartid. Hence, this DT node is required only for booting Linux kernel as EFI application. For other OS which don't support booting as EFI application, passing boot_hartid via DT node is not required. Correct? If that is the case, isn't it easier use volatile UEFI variable which is set by the EDK2/u-boot and read by the kernel if DT is not there?

Normal kernels should be using EFI, either directly (the Linux EFI stub case) or indirectly (a bootloader like GRUB or FreeBSD's loader). SBI booting is legacy and should not be used, but Linux is behind other operating systems like FreeBSD in actually using EFI (which has been doing so for around 18 months at this point), I assume because GRUB still doesn't work properly.

> Any S-mode OS following the traditional RISC-V booting protocol don't need the boot-hartid DT property.

Such OSes should stop existing as time goes on, migrating to EFI.

> Only EFI stub based booting has no provision for passing boot-hartid so the boot-hartid DT property is a temporary work-around for EFI booting and this will break for UEFI+ACPI systems which do not have DT.

"EFI stub" is a Linux-specific thing, and a specific form of Linux EFI booting. Linux also boots via EFI using GRUB, and non-Linux OSes like FreeBSD don't have an equivalent, they have a real separate bootloader they ship with, both of which are affected in just the same way.

**Comment by atishp04** [comment]
> > After the last M-mode booting stage (OpenSBI), only boot HART continues booting. The hartid of the booting hart is passed via a0 register. It is responsibility of S-mode software to save the hartid in some per-CPU structure.
> > Based on above approach, most S-mode software (such as, U-boot S-mode, Linux, Xvisor, etc) don't require special SBI extension for knowing boot hartid.
> 
> _Most_ S-mode software should be booting via EFI, not directly from SBI firmware.
> 

I don't think anybody is suggesting to booting directly from SBI firmware. While EFI booting is mandated in the platform spec, other booting methods from U-Boot (booti, bootm) should also be supported always. 

> > The boot-hartid being passed in DT by EDK2 is a work-around for EFI booting protocol. @atishp04 was there any other reason for passing boot-hartid via DT?
> 
> I'm talking about how U-Boot behaves today when booting via EFI, not what the nascent EDK2 port does.
> 
> > If I understand correctly, normal kernel should expect a0 to have the boot_hartid. Hence, this DT node is required only for booting Linux kernel as EFI application. For other OS which don't support booting as EFI application, passing boot_hartid via DT node is not required. Correct? If that is the case, isn't it easier use volatile UEFI variable which is set by the EDK2/u-boot and read by the kernel if DT is not there?
> 
> Normal kernels should be using EFI, either directly (the Linux EFI stub case) or indirectly (a bootloader like GRUB or FreeBSD's loader). SBI booting is legacy and should not be used, but Linux is behind other operating systems like FreeBSD in actually using EFI (which has been doing so for around 18 months at this point), I assume because GRUB still doesn't work properly.
> 
> > Any S-mode OS following the traditional RISC-V booting protocol don't need the boot-hartid DT property.
> 
> Such OSes should stop existing as time goes on, migrating to EFI.

Again, EFI should be preferred standard(to support distros) but not the only method for booting.

> 
> > Only EFI stub based booting has no provision for passing boot-hartid so the boot-hartid DT property is a temporary work-around for EFI booting and this will break for UEFI+ACPI systems which do not have DT.
> 
> "EFI stub" is a Linux-specific thing, and a specific form of Linux EFI booting. Linux also boots via EFI using GRUB, and non-Linux OSes like FreeBSD don't have an equivalent, they have a real separate bootloader they ship with, both of which are affected in just the same way.

**Comment by atishp04** [comment]
@vlsunil We can explore the RISC-V specific EFI protocol route suggested by Ard.
But I am just curios how much effort is to add small smbios table parsing support in EFI stub ?
The UEFI spec already mandates that. So you need to setup the smbios table for boothartid anyways.

UEFI firmware must exposes RISC-V boot hart ID in the firmware data structure to OS:
• If platform supports SMBIOS, then SMBIOS type 44 record for the boot hart must have “Boot 
Hart” set to 1 in “RISC-V Processor-specific Data” structure.
• If platform supports Device Tree, the Device Tree must contains a unsigned integer (32bit) 
property "boot- hartid" under the /chosen node which would indicate the booting hart ID to 
the supervisor OS.

@jrtc27 Is that an acceptable solution for FreeBSD ?

**Comment by jrtc27** [comment]
It's a pain, and SMBIOS is not mandatory in UEFI. It's a really heavyweight way of getting the current hartid, which you need before you can even use the PLIC (or ACLINT, now there's S-mode-visible devices there) since you need to know which target/context you are. Similarly all the extension, including MMU, information is tied to your hart in the DT or ACPI tables, so it's not just about knowing your hartid by the time you want to bring up other cores, it's needed for all manner of initialisation.

I don't see why this needs to be so complicated. Just define a trivial SBI "get current hart id" extension (which may have no physical meaning, just corresponds to whatever's in the device tree or ACPI tables for the cpu nodes, and what would be passed in a0 if booting directly from SBI, but most implementations will implement it trivially as a read of mhartid), that then supports all boot methods, be it DT or ACPI and EFI via U-Boot, EFI via EDK2 or some other firmware-specific method, and can be used immediately in the kernel.

**Comment by jrtc27** [comment]
Well, except you've gone and made it mandatory for RISC-V UEFI if using ACPI, which I don't think is a normal requirement to have... is that true of AArch64?

**Comment by jrtc27** [comment]
A RISC-V-specific EFI protocol would also be fine by me though, I just think that this is something that should be an SBI extension, just like you can already query things like mvendorid, as it's all just papering over ISA deficiencies and thus what SBI is used for half the time.

**Comment by atishp04** [comment]
> A RISC-V-specific EFI protocol would also be fine by me though, I just think that this is something that should be an SBI extension, just like you can already query things like mvendorid, as it's all just papering over ISA deficiencies and thus what SBI is used for half the time.

New SBI extension should be added only if there is no other way to solve a generic problem such as ISA gaps. The boot hartid issue is very specific to efi booting only. Any system that doesn't require EFI booting won't need it. Even for EFI booting, we have other approaches such EFI protocol that can solve the problem. As it is a EFI specific problem, it should be solved in EFI specific way if possible. There is no other way to get the vendorid other than a SBI call. SBI extension should be the last resort rather than first.

**Comment by atishp04** [comment]
> Well, except you've gone and made it mandatory for RISC-V UEFI if using ACPI, which I don't think is a normal requirement to have... is that true of AArch64?
Do you mean SMBIOS is mandated for RISC-V UEFI if ACPI is used ? If yes, UEFI spec doesn't mandate SMBIOS. It says that "**if SMBIOS is supported**" then SMBIOS type 44 will be used for boot hart.
And I did not propose the ECR for this text or involved in the draft w.r.t SMBIOS. Abner may have better idea about the reasoning for SMBIOS route.

**Comment by jrtc27** [comment]
> > A RISC-V-specific EFI protocol would also be fine by me though, I just think that this is something that should be an SBI extension, just like you can already query things like mvendorid, as it's all just papering over ISA deficiencies and thus what SBI is used for half the time.
> 
> New SBI extension should be added only if there is no other way to solve a generic problem such as ISA gaps. The boot hartid issue is very specific to efi booting only. Any system that doesn't require EFI booting won't need it.

I really do not care about such systems, we should do as much as we can to discourage their existence. If you're running an S-mode OS, you should use EFI, and it is a waste of time to keep talking about such bad practices.

> Even for EFI booting, we have other approaches such EFI protocol that can solve the problem. As it is a EFI specific problem, it should be solved in EFI specific way if possible. There is no other way to get the vendorid other than a SBI call. SBI extension should be the last resort rather than first.

That's one way of looking at it. I take the view that SBI is for fixing things that should have been in the ISA, like shartid, and EFI is for providing higher-level bootloader and OS services. But as I say, I'm totally fine with adding a simple EFI protocol, just not my personal first choice.

**Comment by jrtc27** [comment]
> > Well, except you've gone and made it mandatory for RISC-V UEFI if using ACPI, which I don't think is a normal requirement to have... is that true of AArch64?
>
> Do you mean SMBIOS is mandated for RISC-V UEFI if ACPI is used ? If yes, UEFI spec doesn't mandate SMBIOS. It says that "**if SMBIOS is supported**" then SMBIOS type 44 will be used for boot hart.
> And I did not propose the ECR for this text or involved in the draft w.r.t SMBIOS. Abner may have better idea about the reasoning for SMBIOS route.

Well, it says "UEFI firmware must exposes RISC-V boot hart ID in the firmware data structure to OS" and then lists two ways to do that, which I take as meaning "you must do at least one of these". Either it imposes SMBIOS if you're not providing a device tree, which is overly restrictive, or it isn't a complete list, which needs fixing.

Who actually reviewed what ended up in the UEFI specs, anyway? I find it concerning that this kind of thing is in there already, surely it should've gone through public review and been signed off by RISC-V International first given it's an integral part of RISC-V platforms?

**Comment by atishp04** [comment]
> > > A RISC-V-specific EFI protocol would also be fine by me though, I just think that this is something that should be an SBI extension, just like you can already query things like mvendorid, as it's all just papering over ISA deficiencies and thus what SBI is used for half the time.
> > 
> > 
> > New SBI extension should be added only if there is no other way to solve a generic problem such as ISA gaps. The boot hartid issue is very specific to efi booting only. Any system that doesn't require EFI booting won't need it.
> 
> I really do not care about such systems, we should do as much as we can to discourage their existence. If you're running an S-mode OS, you should use EFI, and it is a waste of time to keep talking about such bad practices.
> 

Well, there another side which doesn't prefer EFI booting at all. We received many objections why it was mandated in the platform spec as well. It was clearly communicated that EFI is the only standard booting method all the distros support. Thus, EFI booting must be mandatory. However, we should not discourage anybody in doing something else (U-boot) if they wish to do so. 


> > Even for EFI booting, we have other approaches such EFI protocol that can solve the problem. As it is a EFI specific problem, it should be solved in EFI specific way if possible. There is no other way to get the vendorid other than a SBI call. SBI extension should be the last resort rather than first.
> 
> That's one way of looking at it. I take the view that SBI is for fixing things that should have been in the ISA, like shartid, and EFI is for providing higher-level bootloader and OS services. But as I say, I'm totally fine with adding a simple EFI protocol, just not my personal first choice.

Thanks! We can continue discussing the EFI protocol details in the linux mailing list with EFI experts.

**Comment by atishp04** [comment]
> > > Well, except you've gone and made it mandatory for RISC-V UEFI if using ACPI, which I don't think is a normal requirement to have... is that true of AArch64?
> > 
> > 
> > Do you mean SMBIOS is mandated for RISC-V UEFI if ACPI is used ? If yes, UEFI spec doesn't mandate SMBIOS. It says that "**if SMBIOS is supported**" then SMBIOS type 44 will be used for boot hart.
> > And I did not propose the ECR for this text or involved in the draft w.r.t SMBIOS. Abner may have better idea about the reasoning for SMBIOS route.
> 
> Well, it says "UEFI firmware must exposes RISC-V boot hart ID in the firmware data structure to OS" and then lists two ways to do that, which I take as meaning "you must do at least one of these". Either it imposes SMBIOS if you're not providing a device tree, which is overly restrictive, or it isn't a complete list, which needs fixing.

It's the latter. It's not the complete list. Once we have a EFI protocol in place, it will probably added to the list.

> 
> Who actually reviewed what ended up in the UEFI specs, anyway? I find it concerning that this kind of thing is in there already, surely it should've gone through public review and been signed off by RISC-V International first given it's an integral part of RISC-V platforms?

AFAIK, it was discussed in UEFI forum and got approved. @changab can throw some light on the ECR approval process and SMBIOS part.

**Comment by jrtc27** [comment]
> > > > Well, except you've gone and made it mandatory for RISC-V UEFI if using ACPI, which I don't think is a normal requirement to have... is that true of AArch64?
> > > 
> > > 
> > > Do you mean SMBIOS is mandated for RISC-V UEFI if ACPI is used ? If yes, UEFI spec doesn't mandate SMBIOS. It says that "**if SMBIOS is supported**" then SMBIOS type 44 will be used for boot hart.
> > > And I did not propose the ECR for this text or involved in the draft w.r.t SMBIOS. Abner may have better idea about the reasoning for SMBIOS route.
> > 
> > 
> > Well, it says "UEFI firmware must exposes RISC-V boot hart ID in the firmware data structure to OS" and then lists two ways to do that, which I take as meaning "you must do at least one of these". Either it imposes SMBIOS if you're not providing a device tree, which is overly restrictive, or it isn't a complete list, which needs fixing.
> 
> It's the latter. It's not the complete list. Once we have a EFI protocol in place, it will probably added to the list.

Once it's added, I think it should become mandatory. SMBIOS can be removed given it's never existed (except maybe in QEMU prototypes? though I didn't even know riscv-smbios existed as a document until today, and it has no declaration of its lifecycle status, just self-declared as v1.0; again, this needs proper community feedback, review and ratification like any other platform spec) and is obsolete, and /chosen/boot-hartid can be left as legacy "if using a device tree you must also put a copy of the boot hartid here" (maybe removed at some point in the future, I don't know what the UEFI forum's view on backwards compatibility is).

> > Who actually reviewed what ended up in the UEFI specs, anyway? I find it concerning that this kind of thing is in there already, surely it should've gone through public review and been signed off by RISC-V International first given it's an integral part of RISC-V platforms?
> 
> AFAIK, it was discussed in UEFI forum and got approved. @changab can throw some light on the ECR approval process and SMBIOS part.

I don't really care about them, I care about getting feedback from the people who actually have to use the thing (OS developers, and not just what one or two Linux developers thing) and getting review by the RISC-V community due to how important it is in the ecosystem, just like the SBI spec, ABI spec, platforms spec, etc. Side-stepping the issue leads to external organisations specifying things for key RISC-V platform components that isn't how the RISC-V community or RISC-V International wants them, both in terms of functionality and document quality (the former crucial, the latter "just" reflects poorly on RISC-V).

**Comment by changab** [comment]
> AFAIK, it was discussed in UEFI forum and got approved. @changab can throw some light on the ECR approval process and SMBIOS part.

We reviewed this ECR in the email thread and then submitted it to USWG then it got approved, that's it.
We added the boot hart ID in the structure for the later boot processes. In DT for the DT-aware SW/FW and in SMBIOS for the SMBIOS-aware SW/FW. Both for the boot loader, OS, or the pre-OS environment such as EFI shell for the diag, factory processes, or other EFI applications that require this information. We can add the EFI solution if it is defined later. Not only the boot hart ID information but also the number of harts and the corresponding hart ID,  the information of supported ISA and non-ISA RISC-V processor features that translated from configuration structure. The above information may be needed for the software after FW booting the system.

**Comment by vlsunil** [comment]
> @vlsunil We can explore the RISC-V specific EFI protocol route suggested by Ard. But I am just curios how much effort is to add small smbios table parsing support in EFI stub ? The UEFI spec already mandates that. So you need to setup the smbios table for boothartid anyways.
> 
@atishp04 - Parsing SMBIOS early will have similar challenges. It uses memory management services to do ioremap etc. We could probably do some hack and read it early before VM initialization but I think that is a pain and not a clean solution. I doubt people will agree for such thing. Also, SMBIOS should not be required for OS booting or core kernel functions IMHO.

Is this calling convention of passing boot hart id in a0 is documented somewhere so that OSVs and FW vendors (considering any future OS) understand the requirement?

**Comment by changab** [comment]
> SMBIOS should not be required for OS booting or core kernel functions IMHO.

Same thought here, UEFI spec only says "If platform supports SMBIOS...", such as the Linux OS we ship with our servers that support SMBIOS.

@vlsunil This is a long thread and I probably miss that somewhere. We can't parse DT to get the boot hart ID in EFI Stub?

**Comment by avpatel** [comment]
@vlsunil The SMBIOS address in EFI tables is a physical address. We can easily access data from SMBIOS without mapping anything.

The RISC-V booting protocol of having hartid in a0 register and boot parameter in a1 register is being followed since early days of RISCV (more than 10 years). In fact, SBI HSM extension has also adopted same convention for HART hotplug. The rationale is that S-mode software can save the hartid in some per-hart context and access it whenever required. This works perfectly fine for all S-mode OSes and hypervisor.

Only EFI booting protocol is different so we have to find a solution which is specific to EFI rather. Adding boot-hartid in SMBIOS looks like good solution which is independent of DT and ACPI. Also, the platform spec can always mandate SMBIOS for systems with UEFI.

Regards,
Anup

**Comment by jrtc27** [comment]
It can but it shouldn't. ACPI+UEFI without SMBIOS (be it its presence and/or use by the OS) is totally viable on AArch64, and something we routinely use, so we should not be making it impossible for that to work on RISC-V because of a single integer that needs to be passed to the OS somehow.

**Comment by jrtc27** [comment]
i.e. just define an EFI protocol or SBI extension and move on

**Comment by avpatel** [comment]
> > @vlsunil We can explore the RISC-V specific EFI protocol route suggested by Ard. But I am just curios how much effort is to add small smbios table parsing support in EFI stub ? The UEFI spec already mandates that. So you need to setup the smbios table for boothartid anyways.
> > 
> @atishp04 - Parsing SMBIOS early will have similar challenges. It uses memory management services to do ioremap etc. We could probably do some hack and read it early before VM initialization but I think that is a pain and not a clean solution. I doubt people will agree for such thing. Also, SMBIOS should not be required for OS booting or core kernel functions IMHO.
> 

Can you elaborate why early parsing of SMBIOS will not be clean? The Linux RISC-V kernel already does early parsing of DT for memblocks and later doing full parsing.

Both EFI tables and SMBIOS are well defined structures so any part of kernel should be free to access it early.

Regards,
Anup

**Comment by kumarsankaran** [comment]
@avpatel My suggestion is that the OS should not be dependent on an SMBIOS table to be present for booting. Agree with @jrtc27 that systems support UEFI + ACPI without the SMBIOS table and this is a valid configuration.

**Comment by jrtc27** [comment]
> > > @vlsunil We can explore the RISC-V specific EFI protocol route suggested by Ard. But I am just curios how much effort is to add small smbios table parsing support in EFI stub ? The UEFI spec already mandates that. So you need to setup the smbios table for boothartid anyways.
> > 
> > 
> > @atishp04 - Parsing SMBIOS early will have similar challenges. It uses memory management services to do ioremap etc. We could probably do some hack and read it early before VM initialization but I think that is a pain and not a clean solution. I doubt people will agree for such thing. Also, SMBIOS should not be required for OS booting or core kernel functions IMHO.
> 
> Can you elaborate why early parsing of SMBIOS will not be clean? The Linux RISC-V kernel already does early parsing of DT for memblocks and later doing full parsing.
> 
> Both EFI tables and SMBIOS are well defined structures so any part of kernel should be free to access it early.

No clue about Linux, but on FreeBSD SMBIOS is treated as a normal driver that attaches at the same time as any other. There's some special stuff in the bootloader just for x86 that also pokes at SMBIOS tables and passes that to the kernel, but it's to deal with x86 historical ugliness.

**Comment by avpatel** [comment]
> 

@jrtc27 Can you give an example of a potential EFI protocol for RISC-V ? Is it possible to say that certain CSR/Register need to be set with boot-hartid before jumping to EFI application ?

**Comment by avpatel** [comment]
> @avpatel My suggestion is that the OS should not be dependent on an SMBIOS table to be present for booting. Agree with @jrtc27 that systems support UEFI + ACPI without the SMBIOS table and this is a valid configuration.

An alternate proposal is to extend the ACPI FADT. In ARM kernel, the PSCI needs to know what kind of instruction to use (SMC/HVC) so for these they have added ARM_BOOT_ARCH field in ACPI FADT. We can do something similar for RISC-V and add 32/64-bit RISCV_BOOT_ARCH field in FADT where the lower bits will be boot-hartid.

Now when booting as EFI application, if DT is provided then boot-hartid is obtained from /chosen node whereas if ACPI is provided then boot-hartid is obtained from RISCV_BOOT_ARCH field in ACPI FADT.

Anyways, ACPI for RISC-V is work-in-progress so it should be possible to define RISCV_BOOT_ARCH for FADT now.

Regards,
Anup

**Comment by vlsunil** [comment]
> Can you elaborate why early parsing of SMBIOS will not be clean? The Linux RISC-V kernel already does early parsing of DT for memblocks and later doing full parsing.
> 
> Both EFI tables and SMBIOS are well defined structures so any part of kernel should be free to access it early.

May be we can. But I think we can't reuse the existing initialization and parsing code which assumes there is no identity mapping for these addresses and sets up mapping before accessing. 
Implementing such early parsers and maintaining looks like heavy lifting to me compared to other solutions possible. Logically also, it may be better to restrict EFI stub to EFI spec related instead of overloading with these FW tables.

I think Ard's suggestion to implement a new protocol like TCG defined (https://trustedcomputinggroup.org/wp-content/uploads/EFI-Protocol-Specification-rev13-160330final.pdf). This is something similar to variable approach with RVI specific GUID. I prefer this approach than trying to use SMBIOS or ACPI early. 

ARM_BOOT_ARCH is also parsed late after ACPI tables are initialized. The issue is with this very early part of the code.

**Comment by avpatel** [comment]
I am fine with a EFI RISCV Boot Hartid protocol (just like Ard suggested).

Regarding ACPI FADT approach, we can always parse boot-hartid early before full ACPI parsing is done by generic ACPI support.

Regards,
Anup

**Comment by changab** [comment]
> SMBIOS can be removed given it's never existed (except maybe in QEMU prototypes?

@jrtc27 EDK2 RISC-V port has the implementation.

**Comment by changab** [comment]
> d it has no declaration of its lifecycle status, just self-declared as v1.0; again, this needs proper community feedback, review and ratification like any other platform spec

@jrtc27 
Thanks for the reminder. I don't pay much attention to RISC-V SMBIOS spec recently and didn't touch the content for a while/ Because only edk2 has the implementation and most of the current RISC-V platforms are not using edk2 as the FW solution. I will add "not ratified" to the spec title for now.

**Comment by avpatel** [comment]
Atish already summarized past discussions on LKML but here's another view more in-context of new SBI extension.
(Refer, https://lore.kernel.org/all/c758bae4-3578-f062-edfd-f5835ae525fe@gmx.de/T/#me557d1b422b402f45a8be221eadcc5f1125d4290)

Some of the downsides of SBI extension returning boot-hartid/hartid are as follows:
1. This will be used only once when kernel is booted as EFI application to determince boot-hartid and it will never be used afterwards at runtime
2. Kernel booted using non-EFI method will never use it because hartid is available in a0 register
3. We also have hypervisors (such as Xen) which prefer booting kernel directly without any bootloader so this SBI extension will be redundant for such hypervisors

My concern is largely around utility value of a SBI extension returning boot-hartid/hartid. The boot-hartid is a transient read-only information which is not of much use at runtime and this can be easily part of boot protocol (DT/UEFI/ACPI). The DT and ACPI will anyway list all HARTs with their corresponding hartids so SBI extension returning hartid is not of much use as well.

Recently, we dropped the SBI DMA Sync extension proposal (which I had proposed) in-favour of CMO trap-n-emulate because SBI DMA Sync extension will not be preferred by S-mode software whenever it has both SBI DMA sync and CMO instructions available. Luckily, RVI ratified CMO instructions this year so there is no need to take SBI DMA Sync forward.

Generally speaking, we define a new SBI extension only if:
1. the SBI extension has good utility value across bootloaders, OSes, and hypervisors
2. the SBI extension is not competing with any of the ISA extension or non-ISA specification

Unfortunately, criteria/policy for new SBI extensions is not written anywhere and someday we will have to define a policy for new SBI extensions.

Regards,
Anup
